### PR TITLE
Finish typescripting chat.js

### PR DESCRIPTION
**Pull request by CheeseMuffin**
This currently has the minimum requirements to make TypeScript happy for chat.js; most of the function parameters are undocumented. Let me know if you would like me to (to the best of my ability) add descriptions for them.

### CODE REVIEW

File: chat.js [hunk 1]
```diff
@@ -394,11 +408,17 @@ class CommandContext {
...
 		return commandHandler;
 	}
+	/**
+	 * @param {string | {call: Function}} commandHandler
+	 */
 	run(commandHandler) {
+		let realCommandHandler = typeof commandHandler === 'string' ?
+			                          Chat.commands[commandHandler] :
+			                          commandHandler;
 		if (typeof commandHandler === 'string') commandHandler = Chat.commands[commandHandler];
```
Comment (KrisOnline) [inline chat.js]: isnt this unnecessary now?

File: chat.js [hunk 2]
```diff
@@ -394,11 +408,17 @@ class CommandContext {
 
 		return commandHandler;
 	}
+	/**
+	 * @param {string | {call: Function}} commandHandler
+	 */
 	run(commandHandler) {
+		let realCommandHandler = typeof commandHandler === 'string' ?
+			                          Chat.commands[commandHandler] :
+			                          commandHandler;
```
Comment (Zarel) [inline chat.js]: Would rather you just `// @ts-ignore` instead of rewriting code here.

File: chat.js [hunk 3]
```diff
@@ -753,20 +849,20 @@ class CommandContext {
 				}
 			}
 
-			if (!this.checkFormat(room, user, message)) {
+			if (room && !this.checkFormat(room, user, message)) {
```
Comment (Zarel) [inline chat.js]: Update `checkFormat`'s signature to `room?` instead of doing this here.

File: chat.js [hunk 4]
```diff
@@ -753,20 +849,20 @@ class CommandContext {
...
 				return false;
 			}
 
-			if (!this.checkSlowchat(room, user) && !user.can('mute', null, room)) {
+			if (room && !this.checkSlowchat(room, user) && !user.can('mute', null, room)) {
```
Comment (Zarel) [inline chat.js]: Ditto this and banWords

File: chat.js [hunk 5]
```diff
@@ -961,11 +1078,16 @@ Chat.CommandContext = CommandContext;
  */
 Chat.parse = function (message, room, user, connection) {
 	Chat.loadPlugins();
-	let context = new CommandContext({message, room, user, connection});
+	let context = new CommandContext({message: message, room: room, user: user, connection: connection});
```
Comment (Zarel) [inline chat.js]: This should not be necessary...

File: chat.js [hunk 6]
```diff
@@ -1082,7 +1208,7 @@ Chat.html = function (strings, ...args) {
  * @param  {any} num
  * @param  {?string} plural
  * @param  {?string} singular
- * @return {string}
+ * @return {?string}
```
Comment (Zarel) [inline chat.js]: This should be `{string}`

File: chat.js [hunk 7]
```diff
@@ -1082,7 +1208,7 @@ Chat.html = function (strings, ...args) {
  * @param  {any} num
  * @param  {?string} plural
  * @param  {?string} singular
```
Comment (Zarel) [inline chat.js]: This should be `{string} [singular]`, ditto `plural`

File: chat.js [hunk 8]
```diff
@@ -1106,13 +1232,16 @@ Chat.plural = function (num, plural = 's', singular = '') {
  */
 Chat.toTimestamp = function (date, options) {
 	const human = options && options.human;
+	/** @type {Array<number>} */
```
Comment (Zarel) [inline chat.js]: PS standard is to match TypeScript standard: `number[]` rather than `Array<number>`

Comment (Zarel) [reply chat.js]: Also, I'd use `any[]`. Don't worsen optimization just to make TypeScript happy.

File: package-lock.json [hunk 9]
```diff
@@ -1785,7 +1785,8 @@
     "permessage-deflate": {
       "version": "0.1.6",
       "resolved": "https://registry.npmjs.org/permessage-deflate/-/permessage-deflate-0.1.6.tgz",
-      "integrity": "sha1-WB8c7fvUQPrEfQd3vohjM4a5kt4="
+      "integrity": "sha1-WB8c7fvUQPrEfQd3vohjM4a5kt4=",
+      "optional": true
```
Comment (Zarel) [inline package-lock.json]: Revert your copy of `package-lock.json`

File: chat.js [hunk 10]
```diff
@@ -766,6 +861,7 @@ class CommandContext {
 				this.errorReply(`Your username contains a phrase banned by this room.`);
 				return false;
 			}
+
```
Comment (Zarel) [inline chat.js]: can you reverse this?

File: rooms.js [hunk 11]
```diff
@@ -1004,6 +1006,12 @@ class BasicChatRoom extends BasicRoom {
 		/** @type {Map<string, ChatRoom>?} */
 		this.subRooms = null;
 
+		/** @type {?true | RegExp} */
+		this.banwordRegex = null;
+
+		/** @type {Array<String>} */
```
Comment (Zarel) [inline rooms.js]: `string[]`

**Comment by Zarel** [comment]
Everything else looks good! Thanks!

**Comment by CheeseMuffin** [comment]
hm, I switched to a new timezone yesterday, but not sure why that would be messing up the timestamps here (time on my computer seems correct).

**Comment by Zarel** [comment]
Thanks!

**Pull request merged**